src/libopensc/reader-pcsc.c

```diff
@@ -378,7 +378,10 @@ static int refresh_attributes(sc_reader_t *reader)
 		if (rv == (LONG)SCARD_E_TIMEOUT) {
 			/* Timeout, no change from previous recorded state. Make sure that
 			 * changed flag is not set. */
-			reader->flags &= ~SC_READER_CARD_CHANGED;
+			if (priv->reader_state.dwEventState & SCARD_STATE_PRESENT && !(reader->flags & SC_READER_CARD_PRESENT))
+				reader->flags |= SC_READER_CARD_CHANGED;
```

Comment:
Good point, I will update the comment.
Regarding the `SCardGetStatusChange()`, when no change happens, the `SCARD_E_TIMEOUT` is returned. Then, when reader is removed, next call returns `SCARD_E_UNKNOWN_READER` and `SC_READER_CARD_PRESENT` is removed from flags (but no `SC_READER_CARD_CHANGED` is set though). When reader is then inserted back again, the return code is again `SCARD_E_TIMEOUT`.

Reply:
The comment above (and the change to `refresh_attributes`) is somewhat confusing. If a reader is removed, then it should be treated as removed. If it is inserted again, it should be treated like a new reader, because all card state is lost and additionally you can't be sure whether this is really the very same card that's inserted.

For me this seems like `SCARD_E_UNKNOWN_READER` is not treated correctly. This change may help (note that the first change actually fixes a check to `SC_READER_REMOVED` as well): 
```diff
diff --git a/src/libopensc/reader-pcsc.c b/src/libopensc/reader-pcsc.c
index 04d5ac8f..8c2f98c9 100644
--- a/src/libopensc/reader-pcsc.c
+++ b/src/libopensc/reader-pcsc.c
@@ -364,7 +364,7 @@ static int refresh_attributes(sc_reader_t *reader)
        if (reader->ctx->flags & SC_CTX_FLAG_TERMINATE)
                return SC_ERROR_NOT_ALLOWED;
 
-       if (priv->reader_state.szReader == NULL || reader->ctx->flags & SC_READER_REMOVED) {
+       if (priv->reader_state.szReader == NULL || reader->flags & SC_READER_REMOVED) {
                priv->reader_state.szReader = reader->name;
                priv->reader_state.dwCurrentState = SCARD_STATE_UNAWARE;
                priv->reader_state.dwEventState = SCARD_STATE_UNAWARE;
@@ -394,6 +394,8 @@ static int refresh_attributes(sc_reader_t *reader)
 #endif
                                || rv == (LONG)SCARD_E_SERVICE_STOPPED) {
                        reader->flags &= ~(SC_READER_CARD_PRESENT);
+                       reader->flags |= SC_READER_REMOVED;
+                       priv->gpriv->removed_reader = reader;
                        SC_FUNC_RETURN(reader->ctx, SC_LOG_DEBUG_VERBOSE, SC_SUCCESS);
                }
```